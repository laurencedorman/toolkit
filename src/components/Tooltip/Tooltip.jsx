/**
 * Copyright (c) Colibri SAS - ManoMano
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

/* eslint-disable */
// TODO in https://manomano.atlassian.net/browse/REACT-240

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Transition, animated } from 'react-spring';
import cn from 'classnames';

import Portal from '../Portal';
import Toggle from '../Toggle';
import tooltipPosition from './tooltipPosition';

import styles from './Tooltip.module.scss';

let addTooltipPosition = {};

const renderTooltip = (display, content, bgColor) => {
  return (
    <Portal>
      <Transition
        native
        unique
        config={{ tension: 350, friction: 23, mass: 0.6, precision: 0.1 }}
        items={display}
        from={{ o: 0, s: 0.6 }}
        enter={{ o: 1, s: 1 }}
        leave={{ o: 0, s: 0.6 }}
      >
        {displayTooltip =>
          displayTooltip
            ? // eslint-disable-next-line react/prop-types,react/display-name
              ({ o, s }) => (
                <div
                  className={styles.tooltip}
                  style={addTooltipPosition.style}
                >
                  <div className={styles[addTooltipPosition.class]}>
                    <animated.div
                      style={{
                        opacity: o.interpolate(opacity => opacity),
                        transform: s.interpolate(scale => `scale(${scale})`),
                        backgroundColor: bgColor,
                        color: '#fff',
                      }}
                    >
                      <div className={styles.svg}>
                        <svg
                          xmlns="http://www.w3.org/2000/svg"
                          viewBox="0 0 18.39 7.96"
                          width="16"
                          height="16"
                        >
                          <path
                            d="M18.39,0,11.31,7.08a3,3,0,0,1-4.23,0L0,0Z"
                            fill={bgColor}
                          />
                        </svg>
                      </div>
                      <div className={styles.content}>
                        <span className={styles.spanContent}>{content}</span>
                      </div>
                    </animated.div>
                  </div>
                </div>
              )
            : () => null
        }
      </Transition>
    </Portal>
  );
};

export default class Tooltip extends PureComponent {
  constructor(props) {
    super(props);
    this.transmitter = React.createRef();
  }

  handleMouseEnter = (e, toggleIn) => {
    const { position } = this.props;
    if (this.transmitter.current) {
      addTooltipPosition = tooltipPosition(
        this.transmitter.current.getBoundingClientRect(),
        position
      );
    }
    toggleIn();
  };

  render() {
    const { children, className, dataQa, content, bgColor } = this.props;

    const classNames = cn('toolkit', styles.transmitter, className);

    return (
      <Toggle>
        {({ on, toggleIn, toggleOut }) => (
          <>
            <span
              className={classNames}
              onMouseEnter={e => this.handleMouseEnter(e, toggleIn)}
              onMouseLeave={toggleOut}
              onWheel={toggleOut}
              ref={this.transmitter}
              role="button"
              data-qa={dataQa}
            >
              {children}
            </span>
            {renderTooltip(on, content, bgColor)}
          </>
        )}
      </Toggle>
    );
  }
}

Tooltip.displayName = 'Tooltip';

Tooltip.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
    PropTypes.element,
  ]).isRequired,
  content: PropTypes.node.isRequired,
  position: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),
  className: PropTypes.string,
  bgColor: PropTypes.string,
  dataQa: PropTypes.string,
};

Tooltip.defaultProps = {
  position: 'top',
  className: '',
  bgColor: '#0c193a',
  dataQa: '',
};
