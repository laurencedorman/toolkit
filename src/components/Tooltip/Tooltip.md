Warning

- It includes react-spring as dependency (31.79KB)

Tooltip:

- <b>Tooltip</b> uses <b>Portal</b> component who need some adding to work. Read Portal section to get more information.
- Tooltip works fine with version of react >= <b>16.4</b>
- content can be render a string or function.

Tooltip default:

```js
<p>
  This is a text with tooltip,
  <b>
    <Tooltip content="Awesome, i'm in Portal so no z-index and overflow conflict, Awesome, i'm in Portal so no z-index and overflow conflict, Awesome, i'm in Portal so no z-index and overflow conflict, Awesome, i'm in Portal so no z-index and overflow conflict">
      hover me
    </Tooltip>
  </b>
</p>
```

Tooltip bottom :

```js
<p>
  This is a text with tooltip,
  <b>
    <Tooltip position="bottom" content="Awesome">
      hover me
    </Tooltip>
  </b>
</p>
```

Tooltip left :

```js
<p>
  <b>
    <Tooltip
      position="left"
      content="Awesome, i'm in Portal so no z-index and overflow conflict"
    >
      hover me
    </Tooltip>
  </b>
</p>
```

Tooltip right :

```js
<p>
  This is a text with tooltip,
  <b>
    <Tooltip
      position="right"
      content="Awesome, i'm in Portal so no z-index and overflow conflict"
    >
      hover me
    </Tooltip>
  </b>
</p>
```

Tooltip bottom :

```js
<p>
  This is a text with tooltip,
  <b>
    <Tooltip
      position="bottom"
      content="Awesome, You can add long sentence,  You can add long sentence,  You can add long sentence,"
    >
      hover me
    </Tooltip>
  </b>
</p>
```
