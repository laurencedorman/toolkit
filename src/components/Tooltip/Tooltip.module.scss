/**
 * Copyright (c) Colibri SAS - ManoMano
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@import '../../styles/import.scss';

$marginArrow: em(-10);

.tooltip {
  position: fixed;
  width: auto;
  z-index: z($layers, tooltip);

  & > div {
    position: relative;

    & > div {
      position: relative;
      min-width: em(100);
      max-width: em(300);
      max-height: em(200);
      padding: em($halfSpace) em($space);
      border-radius: em($radius);
      color: $white !important;
      text-align: center;
      backface-visibility: hidden;
    }

    &.top,
    &.bottom {
      transform: translateX(-50%);
    }
    &.left,
    &.right {
      transform: translateY(-50%);
    }
  }

  .content,
  .spanContent {
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
    font-size: em(14);
    white-space: normal;
    word-break: break-word;
    color: $white;
    opacity: 1;
  }
}

.svg {
  position: absolute;
  display: inline-flex;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  z-index: -1;

  svg {
    position: absolute;
  }

  .top &,
  .bottom & {
    justify-content: center;
  }
  .left &,
  .right & {
    align-items: center;
  }

  .top & {
    align-items: flex-end;
    svg {
      bottom: $marginArrow;
    }
  }

  .bottom & {
    align-items: flex-start;
    svg {
      top: $marginArrow;
      transform: rotate(180deg);
    }
  }

  .left & {
    justify-content: flex-end;
    svg {
      right: $marginArrow;
      transform: rotate(-90deg);
    }
  }

  .right & {
    justify-content: flex-start;
    svg {
      left: $marginArrow;
      transform: rotate(90deg);
    }
  }
}

.transmitter {
  position: relative;
  display: inline-flex;
  font-size: inherit;
  font-weight: inherit;
  cursor: pointer;

  &.extend {
    width: 100%;
    justify-content: center;
  }
}
