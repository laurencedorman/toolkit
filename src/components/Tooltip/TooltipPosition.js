const space = 16;
const requiredSpace = 100;


export const tooltipPosition = (ref, position) => {
  const {
    top, left, right, bottom, width, height,
  } = ref.getBoundingClientRect();

  const leftSpace = left + width / 2 + window.scrollX;

  switch (true) {
    case (window.innerHeight - bottom < requiredSpace) && position === 'bottom':
    case position === 'top' && top > requiredSpace:
      return {
        class: 'top',
        style: {
          bottom: window.innerHeight - top + space,
          left: leftSpace,
        },
      };

    case window.innerWidth - right < requiredSpace:
    case position === 'left' && left > requiredSpace:
      return {
        class: 'left',
        style: {
          top: top + (height / 2),
          right: document.body.clientWidth - left + window.scrollX + space,
        },
      };

    case left < requiredSpace:
    case position === 'right' && left > requiredSpace:
      return {
        class: 'right',
        style: {
          top: top + (height / 2),
          left: right + space,
        },
      };

    case top < requiredSpace:
    case position === 'bottom':
      return {
        class: 'bottom',
        style: {
          top: bottom + (space * 2),
          left: leftSpace,
        },
      };

    default:
      return {
        class: 'top',
        style: {
          bottom: window.innerHeight - top + space,
          left: leftSpace,
        },
      };
  }
};

export default tooltipPosition;
