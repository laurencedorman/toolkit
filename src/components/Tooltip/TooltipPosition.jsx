
export const tooltipPositionCheck = (ref, position) => {
  const { top, left, right, bottom } = ref.getBoundingClientRect();
  const requiredSpace = 100;

  if (window.innerWidth - right < requiredSpace) return 'forceLeft';
  else if (left < requiredSpace) return 'forceRight';
  else if (top < requiredSpace) return 'forceBottom';
  else if (window.innerHeight - bottom < requiredSpace) return 'forceTop';
  else if (position === 'top') return 'top';
  else if (position === 'left') return 'left';
  else if (position === 'right') return 'right';
  else if (position === 'bottom') return 'bottom';

  return null;
};

export const tooltipPosition = (ref, checkPosition) => {
  const { top, left, right, width, height, bottom } = ref.getBoundingClientRect();
  const space = 16;
  const lateralPositionWidth = 176;
  const bottomSpace = document.body.clientHeight - top - window.scrollY;
  const sideSpace = document.body.clientWidth - left + window.scrollX;

  /*console.log(`
    top: ${top} \n
    bottom: ${bottom} \n
    left: ${left} \n
    width: ${width} \n
    right: ${right} \n
    height: ${height}
  `);*/

  const topStyle = {
    bottom: bottomSpace + space,
    left: left + width / 2 + window.scrollX,
  };

  const leftStyle = {
    bottom: bottomSpace - space,
    right: sideSpace + space,
  };

  const rightStyle = {
    bottom: bottomSpace - space,
    right: sideSpace - lateralPositionWidth - width - space,
  };

  const bottomStyle = {
    bottom: bottomSpace - (height * 3),
    left: left + width / 2 + window.scrollX,
  };

  /*switch(position) {
    case 'top': return topStyle;
    case 'left': return leftStyle;
    case 'right': return rightStyle;
    case 'bottom': return bottomStyle;
    default: return topStyle
  }*/

  switch(checkPosition) {
    case 'forceLeft':
    case 'left':
      return leftStyle;
    case 'forceRight':
    case 'right':
      return rightStyle;
    case 'forceBottom':
    case 'bottom':
      return bottomStyle;
    case 'forceTop':
    case 'top':
      return topStyle;
    default: return topStyle
  }
};

export default tooltipPosition;
