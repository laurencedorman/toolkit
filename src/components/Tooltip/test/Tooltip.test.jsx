/**
 * Copyright (c) Colibri SAS - ManoMano
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { mount } from 'enzyme';
import Tooltip from '../Tooltip';

describe('Tooltip', () => {
  const wrapper = mount(<Tooltip content="test">test</Tooltip>);

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render content', () => {
    expect(wrapper.props().content).toEqual('test');
  });

  it('should display tooltip', () => {
    wrapper.setProps({ on: true });
    expect(wrapper.props().on).toBe(true);
  });
});
