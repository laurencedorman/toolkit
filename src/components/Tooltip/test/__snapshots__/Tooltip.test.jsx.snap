// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should match snapshot 1`] = `
<Tooltip
  bgColor="#0c193a"
  className=""
  content="test"
  dataQa=""
  position="top"
>
  <Toggle>
    <span
      className="toolkit transmitter"
      data-qa=""
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onWheel={[Function]}
      role="button"
    >
      test
    </span>
    <Portal>
      <Portal
        containerInfo={
          <div
            class="portal"
          />
        }
      >
        <Transition
          config={
            Object {
              "friction": 23,
              "mass": 0.6,
              "precision": 0.1,
              "tension": 350,
            }
          }
          enter={
            Object {
              "o": 1,
              "s": 1,
            }
          }
          from={
            Object {
              "o": 0,
              "s": 0.6,
            }
          }
          items={false}
          keys={[Function]}
          leave={
            Object {
              "o": 0,
              "s": 0.6,
            }
          }
          native={true}
          reset={false}
          unique={true}
        >
          <ForwardRef
            config={
              Object {
                "friction": 23,
                "mass": 0.6,
                "precision": 0.1,
                "tension": 350,
              }
            }
            delay={0}
            filter={[Function]}
            from={
              Object {
                "o": 0,
                "s": 0.6,
              }
            }
            key="false"
            native={true}
            primitive={[Function]}
            reset={false}
            state="enter"
            states={
              Object {
                "enter": Object {
                  "o": 1,
                  "s": 1,
                },
              }
            }
          >
            <KeyframesImpl
              config={
                Object {
                  "friction": 23,
                  "mass": 0.6,
                  "precision": 0.1,
                  "tension": 350,
                }
              }
              delay={0}
              filter={[Function]}
              forwardRef={null}
              from={
                Object {
                  "o": 0,
                  "s": 0.6,
                }
              }
              native={true}
              primitive={[Function]}
              reset={false}
              state="enter"
              states={
                Object {
                  "enter": Object {
                    "o": 1,
                    "s": 1,
                  },
                }
              }
            >
              <Spring
                config={
                  Object {
                    "friction": 23,
                    "mass": 0.6,
                    "precision": 0.1,
                    "tension": 350,
                  }
                }
                delay={0}
                force={false}
                from={
                  Object {
                    "o": 0,
                    "s": 0.6,
                  }
                }
                immediate={false}
                inject={[Function]}
                native={true}
                onRest={[Function]}
                reset={false}
                to={
                  Object {
                    "o": 1,
                    "s": 1,
                  }
                }
              />
            </KeyframesImpl>
          </ForwardRef>
        </Transition>
      </Portal>
    </Portal>
  </Toggle>
</Tooltip>
`;
